### ISSUE TITLE
Versioned dependency and dependency circles

### ISSUE THREAD

**Issue by Edenhofer** [rmarquis/pacaur#370]
### AUR packages which depend on one specific version of another package won't update in case the pkgver of both packages is increased. ###
This is somewhat hard to describe but let's take letsencrypt [1] as an example: It requeries python2-acme [2] to be installed on the system in the version as the package itself. When updating pacaur tries to update each of them individually, which obviosly results in an error (*error: failed to prepare transaction (could not satisfy dependencies)*), since they both need to be updated at the same time.
Pacman handles this properly and recognizes those special cases, hence I suggest to install all pacakges which pacaur has prepared at the same time and not as soon as just one is ready.

[1] https://aur.archlinux.org/packages/letsencrypt/
[2] https://aur.archlinux.org/packages/python2-acme/

**Comment by rmarquis** [comment]
I'll see what I can do. The installation process of built packages needs to be reworked anyway (see #364).

**Comment by Edenhofer** [comment]
Indeed, this might get tricky, though it should be fixable since the information about which dependency is cyclic *(needs to be installed in parrallel or prior by ignoring the conflict)* and which not *(can be installed as usual)*.

**Comment by rmarquis** [comment]
Care to expand? I'm not sure how I could retrieve that information without adding a lot of complexity to the code, but I might be missing something here.

**Comment by Edenhofer** [comment]
I just came up with two rather crappy solutions:
 * just do it how it is done now and evaluate the error message and accordingly wait for those mentioned packages to finish building
 * use the dependcy information from the pacman database to see which packages depend on one specific version of another and evaluate which package can and which can't be installed the usual way
However I admit that both *solutions* would add some degree of complexity to the code.

**Comment by rmarquis** [comment]
The first solution is clearly a no go. If the dependency solver fails, then the whole operation should stop.

The second one looks better on theory, but is actually hard to implement correctly in practice. Dependencies are solved on a "per-dependency level" basis with the deepest dependencies on the tree installed first. This issue would actually need to completely change the installation process in a awkward, prone to error way that would result in a mess imho, especially when the solver has to take multiple packages installation or upgrade at the same time. 

The real question is actually why do you need to depend on this specific dependency version in the first place. I understand  ```letsencrypt``` requires the latest ```python2-acme``` to work, but Arch is a continuously updated distribution, with only the latest version of packages available and with unsupported partial update. If both packages are updated at the same time in the AUR by their maintainers (which seems to be you in both cases), then the solver will do the rest.

In other words, it seems to me that you're the one that creates this issue with a versioned dependency which doesn't make sense on Arch. If this reasoning is flawed in some way, please add your input. I might have completely missed something here.

**Comment by rmarquis** [comment]
No response, closing as 'INVALID'.

**Comment by Edenhofer** [comment]
Sorry for the long delay.
Your conclusion is correct: I am the maintainer and yes I am causing this *bug*. I have fixed the versioned dependencies for now, though I think this should not be a longterm solution. I also admit that Arch does not support partial upgrade. 
However I don't think that this means one should not follow upstream dependencies. Taking the nvidia package [1] as an example, it depends on nvidia-utils [2] but it must be present in the same version. Pacman handles this accordingly, so pacaur should do it for AUR packages a well.
A proper approach would be to check whether the dep is already installed in the requiered version (using pacman's -T option) or whether it will be installed soon after, if so either delay the installation or install it with pacman's --nodeps option.

[1] https://www.archlinux.org/packages/extra/x86_64/nvidia/
[2] https://www.archlinux.org/packages/extra/x86_64/nvidia-utils/

**Comment by rmarquis** [comment]
True, pacman handles this situation correctly and ideally pacaur would handle it too. I however won't work on a theoretical issue. Do you have any existing AUR packages in mind where this user case really happens?

Using ```--nodeps``` might be a possibility, but again solving this issue is tricky as deps are solved by "level" in the dependency chain, and not by packages. In this regards, this issue is kind of similar to #173.

**Comment by Edenhofer** [comment]
I though letsencrypt was a good enaught example. I just guess many AUR package maintainers are getting used to not using versioned deps. Admittetly this is not a serious problem and is fixable by only a small incorrectness in the PKGBUILD.

**Comment by rmarquis** [comment]
Again, letsencrypt was an issue that should not exist in the context of a rolling distribution such as Arch. Versioned dependencies make rarely sense in AUR (although sometimes they do make sense), and pacaur is built around the idea that everything that should be fixed in the PKGBUILDs should be, well, fixed in the PKGBUILDs.

I am aware of only one AUR packages that needs a versioned dependency which causes a cyclic dependency situation (although I can't remember that package name right now..), but it is a slighly different beast. It needs a package to be built to bootstrap its dependency, before the main package can be built and the first package being removed - a user case so specific that only manual intervention can fix.

**Comment by Edenhofer** [comment]
I would argue that versioned deps make as much sense in the AUR as they do in the official repo.
However you are probably right that for this narrow use case the added complexity would not be worth it.

**Comment by rmarquis** [comment]
If there is an existing real user case where this issue does happen, then I'm interested in trying to fix it.

But as far as I can see, we don't have a real user case scenario here which isn't something that should be fixed in the PKGBUILD - and thus this issue does not exist as far as I'm concerned. If you ever came across such an user case, please comment on this ticket - I'll have a look.
